src/sentry/incidents/logic.py

```diff
@@ -684,24 +685,66 @@ def update_alert_rule(
         updated_fields["threshold_period"] = threshold_period
 
     with transaction.atomic():
-        if query is not None or aggregation is not None or time_window is not None:
-            # TODO: We're assuming only one subscription for the moment
-            subscription = (
-                AlertRuleQuerySubscription.objects.select_related("query_subscription")
-                .get(alert_rule=alert_rule)
-                .query_subscription
-            )
-            # If updating any details of the query, update the Snuba subscription
-            update_snuba_subscription(
-                subscription,
-                query if query is not None else alert_rule.query,
-                aggregation
-                if aggregation is not None
-                else QueryAggregations(alert_rule.aggregation),
-                time_window if time_window else alert_rule.time_window,
+        alert_rule.update(**updated_fields)
+        existing_subs = []
+        if (
+            query is not None
+            or aggregation is not None
+            or time_window is not None
+            or projects is not None
+        ):
+            existing_subs = alert_rule.query_subscriptions.all().select_related("project")
+
+        if projects is not None:
+            existing_project_slugs = {sub.project.slug for sub in existing_subs}
+            # Determine whether we've added any new projects as part of this update
+            new_projects = [
+                project for project in projects if project.slug not in existing_project_slugs
+            ]
+            updated_project_slugs = {project.slug for project in projects}
+            # Find any subscriptions that were removed as part of this update
+            deleted_subs = [
+                sub for sub in existing_subs if sub.project.slug not in updated_project_slugs
+            ]
```

Comment:
If you put the project slugs into sets you could use set operations to do the diff. :bike::house:

Reply:
They are already in sets, but in practice using set operations results in more code, since I then need a mapping of project slug -> (subscription/project)